cmd/kubeadm/app/cmd/join.go

```diff
@@ -162,14 +162,15 @@ func NewCmdJoin(out io.Writer, joinOptions *joinOptions) *cobra.Command {
 			c, err := joinRunner.InitData(args)
 			kubeadmutil.CheckErr(err)
 
+			data := c.(*joinData)
+			fmt.Printf("[join] Using Kubernetes version: %s\n", data.initCfg.KubernetesVersion)
+
 			err = joinRunner.Run(args)
 			kubeadmutil.CheckErr(err)
 
-			// TODO: remove this once we have all phases in place.
-			// the method joinData.Run() itself should be removed too.
-			data := c.(*joinData)
-			err = data.Run()
+			err = showControlPlaneJoinCommand(data, out)
```

Comment:
I think there was a misunderstanding here; please forgive me if I wasn't clear enough in the previous comment

It is correct to move write here the piece of code that writes the final message (as in init), but kubeam join should preserve its own final message.

In other words
1) `showControlPlaneJoinCommand`, `printControlPlaneJoinCommand`, `Tokens` func should go away (because they are not necessary for the kubeadm join final message)
2) the call to `showControlPlaneJoinCommand` should be replaced by the code that outputs the kubeadm join final message (L434-L451 in the original commit). Please note that this code should be refactored a little bit and at the end it should look something similar to
```go
        // if the node is hosting a new control plane instance
	if data.cfg.ControlPlane != nil {
                // outputs the join control plane done message and exit
                etcdMessage := ""	
		if data.cfg.Etcd.External == nil {	
			etcdMessage = "* A new etcd member was added to the local/stacked etcd cluster."	
		}

		ctx := map[string]string{
			"KubeConfigPath": c.KubeConfigPath(),
			"etcdMessage":    etcdMessage,
		}
		joinControPlaneDoneTemp.Execute(j.outputWriter, ctx)
		return nil
	}

	// otherwise, if the node joined as a worker node;
	// outputs the join done message and exit
	fmt.Fprintf(j.outputWriter, joinWorkerNodeDoneMsg)
```

Reply:
done PTAL.